### PR TITLE
We avoid prepending a hardline before a LabelDefinition

### PR DESCRIPTION
**Pull request by Janther**
`LabelDefinition` uses a dedented first line.
This caused a side effect where all Labels would have an empty line before them, this looked a bit quirky but there wasn't anything wrong with it.

After adding the new in depth testing, we recognised that, on the second format, this empty line would be registered as an empty line added by the developer to space out code, and we would preserve this line in `printPreservingEmptyLines`, and we would end up with 2 empty lines instead of one.

This fix might look a bit hacky but it's an edge case given the particular way of printing assembly labels.

### CODE REVIEW

File: tests/Assembly/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -125,6 +125,7 @@ loop:
     a add swap1
     n := sub(n, 1)
     jump(loop)
+
```
Comment (fvictorio) [inline tests/Assembly/__snapshots__/jsfmt.spec.js.snap:128]: Why does this have a space but the `one:`, `two:`, etc. labels below don't?

File: tests_config/run_spec.js [hunk 2]
```diff
@@ -18,17 +18,25 @@ const CURSOR_PLACEHOLDER = '<|>';
 const RANGE_START_PLACEHOLDER = '<<<PRETTIER_RANGE_START>>>';
 const RANGE_END_PLACEHOLDER = '<<<PRETTIER_RANGE_END>>>';
 
-// TODO: these test files need fix
+// Here we add files that will not be the same when formating a second time.
 const unstableTests = new Map(
-  ['Assembly/Assembly.sol'].map((fixture) => {
+  [].map((fixture) => {
```
Comment (mattiaerre) [inline tests_config/run_spec.js:23]: noice now this is empty 😊

File: tests_config/run_spec.js [hunk 3]
```diff
@@ -18,17 +18,25 @@ const CURSOR_PLACEHOLDER = '<|>';
...
       ? fixture
       : [fixture];
     return [path.join(__dirname, '../tests/', file), isUnstable];
   })
 );
 
-const unstableAstTests = new Map();
+// Here we add files that will not have the same AST after being formated.
+const unstableAstTests = new Map(
+  [].map((fixture) => {
```
Comment (mattiaerre) [inline tests_config/run_spec.js:33]: same as above

File: tests/Assembly/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (mattiaerre) [reply tests/Assembly/__snapshots__/jsfmt.spec.js.snap:128]: yeah, they actually got removed

Comment (Janther) [reply tests/Assembly/__snapshots__/jsfmt.spec.js.snap:128]: I added a space here just to have the same code as the example in #449
That way we have an example of no empty line before and 1 empty line before

File: tests_config/run_spec.js (hunk 3 above)
Comment (Janther) [reply tests_config/run_spec.js:33]: I decided to add the function within since an empty `new Map();` doesn't really explain how this map should be implemented in case we had an example where ASTs are changed. 
We don't have inconsistent AST but we do sometimes change it in the parsing step.
Stuff like changing `uint` to `uint256` and enforcing parentheses for easy to read code. For more correctness we could also add a new test that compares the compiled code of contracts where we know that the AST changes.
Until now I have used this [project](https://github.com/Janther/prettier-ast-changing-proposals) to prove that those AST changes don't affect the compiled code, but would be a nice addition to our process and we could keep track on new compiler versions that might affect this.

File: tests/Assembly/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (Janther) [reply tests/Assembly/__snapshots__/jsfmt.spec.js.snap:128]: > yeah, they actually got removed

yeah, before this PR we were always printing an extra line. In the other snapshots, those lines got removed in the output.
This line is from the input of `tests/Assembly/Assembly.sol` so it is added.

**Review by mattiaerre — APPROVED** [review]
good stuff @Janther

File: tests/Assembly/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (fvictorio) [reply tests/Assembly/__snapshots__/jsfmt.spec.js.snap:128]: Oh, sorry :man_facepalming: I shouldn't review code when I'm too tired.

**Review by fvictorio — APPROVED** [review]
